tests/apps/pubsub-subscriber/app.go

```diff
@@ -0,0 +1,178 @@
+}
+
+// this handles /dapr/subscribe, which is called from dapr into this app.
+// this returns the list of topics the app is subscribed to.
+func configureSubscribeHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("configureSubscribeHandler called\n")
+
+	var t topicsList
+	t.Topic = append(t.Topic, pubsubA, pubsubB)
+	log.Printf("configureSubscribeHandler subscribing to:%v\n", t.Topic)
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(t.Topic)
+}
+
+// this handles messages published to "pubsub-a-topic"
+func subscribeHandler(w http.ResponseWriter, r *http.Request) {
+	log.Printf("aHandler is called %s\n", r.URL)
+
+	w.WriteHeader(http.StatusOK)
+	defer r.Body.Close()
+
+	var err error
+	var data []byte
+	var body []byte
+	if r.Body != nil {
+		if data, err = ioutil.ReadAll(r.Body); err == nil {
+			body = data
+			log.Printf("assigned\n")
+		}
+	} else {
+		// error
+		err = errors.New("r.Body is nil")
+	}
+
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		json.NewEncoder(w).Encode(appResponse{
+			Message: err.Error(),
+		})
+		return
+	}
+
+	msg, err := extractMessage(body)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		json.NewEncoder(w).Encode(appResponse{
+			Message: err.Error(),
+		})
+		return
+	}
+
+	lock.Lock()
+	defer lock.Unlock()
+	if strings.HasSuffix(r.URL.String(), pubsubA) {
+		receivedMessagesA = append(receivedMessagesA, msg)
+	} else if strings.HasSuffix(r.URL.String(), pubsubB) {
+		receivedMessagesB = append(receivedMessagesB, msg)
+	} else {
+		w.WriteHeader(http.StatusInternalServerError)
+		json.NewEncoder(w).Encode(appResponse{
+			Message: err.Error(),
+		})
+		return
+	}
+
+	json.NewEncoder(w).Encode(appResponse{
+		Message: "success",
+	})
+}
+
+func extractMessage(body []byte) (string, error) {
+	log.Printf("extractMessage() called")
+
+	log.Printf("body=%s", string(body))
+
+	m := make(map[string]interface{})
+	err := json.Unmarshal(body, &m)
+	if err != nil {
+		log.Printf("Could not unmarshal, %s", err.Error())
+		return "", err
+	}
+
+	msg := m["data"].(string)
+	log.Printf("output='%s'\n", msg)
+
+	return msg, nil
+}
+
+// the test calls this to get the messages received
+func getReceivedMessages(w http.ResponseWriter, r *http.Request) {
+	log.Println("Enter getReceivedMessages")
+
+	response := receivedMessagesResponse{
+		ReceivedByTopicA: receivedMessagesA,
+		ReceivedByTopicB: receivedMessagesB}
+
+	log.Printf("receivedMessagesResponse=%s", response)
+
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(response)
+}
+
+// appRouter initializes restful api router
+func appRouter() *mux.Router {
+	log.Printf("Enter appRouter()")
+	router := mux.NewRouter().StrictSlash(true)
+
+	router.HandleFunc("/", indexHandler).Methods("GET")
+
+	router.HandleFunc("/tests/get", getReceivedMessages).Methods("POST")
+
+	router.HandleFunc("/dapr/subscribe", configureSubscribeHandler).Methods("GET")
```

Comment:
It is odd that a GET method will cause state change. Change it to Post - even if payload is empty.

Reply:
This doesn't cause a state change in the app (it does cause a state change in dapr though).  This matches the [spec](https://github.com/dapr/docs/blob/master/reference/api/pubsub.md#handling-topic-subscriptions):

Reply:
OK. Got it.